[GStreamer][WebRTC] RTP payload type numbers are incorrect and prone to collisions in our SDP offers

philn opened the pull request:
#### 3544d3031a90e44779d593495a9d3eb4ee091047
<pre>
[GStreamer][WebRTC] RTP payload type numbers are incorrect and prone to collisions in our SDP offers
<a href="https://bugs.webkit.org/show_bug.cgi?id=255614">https://bugs.webkit.org/show_bug.cgi?id=255614</a>

Reviewed by Xabier Rodriguez-Calvar.

Use static payload types for the legacy audio codecs, as mandated by their respective RTP payload
RFCs and for dynamic payloads, use unique identifiers, starting at 96. See also
<a href="https://www.iana.org/assignments/rtp-parameters/rtp-parameters.xhtml">https://www.iana.org/assignments/rtp-parameters/rtp-parameters.xhtml</a>

* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::requestPad):
(WebCore::GStreamerMediaEndpoint::pickAvailablePayloadType):
(WebCore::GStreamerMediaEndpoint::createTransceiverBackends):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::payloadTypeForEncodingName):
(WebCore::capsFromRtpCapabilities):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:

Canonical link: <a href="https://commits.webkit.org/263135@main">https://commits.webkit.org/263135@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/14cee52d3f3fb2968c20c4c271637409fd9ce7c4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3768 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3874 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5110 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3967 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3658 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3844 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3773 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3193 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3728 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3958 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3318 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4942 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1464 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3288 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/3356 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3272 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3347 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4715 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3737 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3024 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3271 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3288 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/905 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3303 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3548 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
```diff
@@ -472,6 +472,20 @@ uint32_t UniqueSSRCGenerator::generateSSRC()
+        { "G722"_s, 9 },
+    };
+
+    auto key = String::fromLatin1(encodingName);
```
It would be able to create a `StringView` to avoid copying the `const char*` but I don't think templates are so smart to allow calling the `contains` on that. Maybe you could turn the hash map into `HashMap<StringView, int>` ?

philn replied:
Doesn't work:

```
                                                                                                                                                               
/app/webkit/WebKitBuild/Release/WTF/Headers/wtf/HashMap.h:134:10: note: candidate function not viable: no known conversion from 'const char *' to 'const WTF::HashMap<WTF::StringView, int>::KeyType' (aka 'const WTF
::StringView') for 1st argument                                                                                                                                                                                      
    bool contains(const KeyType&) const;                                                                                                                                                                             
         ^ 
```

philn replied:
Ah but with AtomString it shouldn't require a copy.

philn replied:
Sorry, I don't see how to micro-optimize this. Let's revisit if that really becomes a bottleneck later on. :)

philn committed:
[GStreamer][WebRTC] RTP payload type numbers are incorrect and prone to collisions in our SDP offers
https://bugs.webkit.org/show_bug.cgi?id=255614

Reviewed by Xabier Rodriguez-Calvar.

Use static payload types for the legacy audio codecs, as mandated by their respective RTP payload
RFCs and for dynamic payloads, use unique identifiers, starting at 96. See also
https://www.iana.org/assignments/rtp-parameters/rtp-parameters.xhtml

* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.cpp:
(WebCore::GStreamerMediaEndpoint::requestPad):
(WebCore::GStreamerMediaEndpoint::pickAvailablePayloadType):
(WebCore::GStreamerMediaEndpoint::createTransceiverBackends):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerMediaEndpoint.h:
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.cpp:
(WebCore::payloadTypeForEncodingName):
(WebCore::capsFromRtpCapabilities):
* Source/WebCore/Modules/mediastream/gstreamer/GStreamerWebRTCUtils.h:

Canonical link: https://commits.webkit.org/263135@main
